File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n=len(cost)
        dp=[-1]*(n+1)
        def f(ind):
            if ind==n-1:
                return cost[ind]
            if ind==n-2:
                return cost[ind]
            if dp[ind]!=-1:
                return dp[ind]
            onestep=cost[ind]+f(ind+1)
            twostep=cost[ind]+f(ind+2)
            dp[ind]=min(onestep,twostep)

            return dp[ind]
        return min(f(0),f(1))

File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.cpp
class Solution {
private:
    int f(int i, vector<int>& dp, int n, vector<int>& cost){
        if(i==n-1){
            return cost[i];
        }
        if(i>=n){
            return 0;
        }
        if(dp[i]!=-1){
            return dp[i];
        }
        int onestep=f(i+1,dp,n,cost);
        int twostep=f(i+2,dp,n,cost);

        dp[i]=min(onestep,twostep)+cost[i];
        return dp[i];
    }
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n=cost.size();
        vector<int>dp(n,-1);
        return min(f(0,dp,n,cost),f(1,dp,n,cost));
    }
};